h/models/user.py

```diff
@@ -18,27 +19,74 @@
 PASSWORD_MIN_LENGTH = 2
 
 
+def _normalise_username(username):
+    # We normalize usernames by dots and case in order to discourage attempts
+    # at impersonation.
+    return sa.func.lower(sa.func.replace(username,
+                                         sa.text("'.'"),
+                                         sa.text("''")))
+
+
+class UsernameComparator(Comparator):
+    """
+    Custom comparator for :py:attr:`~h.models.user.User.username`.
+
+    This ensures that all lookups against the username property, such as
+
+        session.query(User).filter_by(username='juanwood')
+
+    use the normalised username for the lookup and appropriately normalise the
+    RHS of the query. This means that a query like the one above will
+    correctly find a user with a username of "Juan.Wood", for example.
+    """
+    def operate(self, op, other):
+        return op(_normalise_username(self.__clause_element__()),
+                  _normalise_username(other))
+
+
 class UserIDComparator(Comparator):
     """
-    Custom comparator for the User.userid property.
+    Custom comparator for :py:attr:`~h.models.user.User.userid`.
+
+    A user's userid is a compound property which depends on their username
+    and their authority. A naive comparator for this property would generate
+    SQL like the following:
+
+        ... WHERE 'acct:' || username || '@' || authority = ...
 
-    This is a comparator conforming to the
-    :py:class:`sqlalchemy.orm.interfaces.PropComparator` interface, which
-    makes it possible to find users by userid without doing ugly SQL
-    concatenation that prevents the use of indices.
+    This would be slow, due to the lack of an index on the LHS expression.
+    While we could add a functional index on this expression, we can also take
+    advantage of the existing index on (normalised_username, authority), which
+    is what this comparator does.
+
+    A query such as
+
+        session.query(User).filter_by(userid='acct:luis.silva@example.com')
+
+    will instead generate
+
+        WHERE
+            (lower(replace(username,     '.', '')), authority    ) =
+            (lower(replace('luis.silva', '.', '')), 'example.com')
     """
     def __init__(self, username, authority):
         self.username = username
         self.authority = authority
 
+    def __clause_element__(self):
+        return sa.tuple_(_normalise_username(self.username), self.authority)
+
     def __eq__(self, other):
```

Comment:
I _think_ I understand what this is now doing, and the different types it will deal with for `other`, but is it worth adding a note here explaining why this has the two different branches (for `string_types` and for everything else)?